implementations/rust/ockam/ockam_api/src/lib.rs

```diff
@@ -595,3 +596,124 @@ pub(crate) fn assert_response_match<'a>(schema: impl Into<Option<&'a str>>, cbor
         }
     }
 }
+
+/// Encode request header and body (if any), send the package to the server and returns its response.
+async fn request<T, R>(
+    ctx: &mut Context,
+    target: &str,
+    label: &str,
+    schema: impl Into<Option<&str>>,
+    route: R,
+    req: &RequestBuilder<'_, T>,
+) -> ockam_core::Result<Vec<u8>>
+where
+    T: Encode<()>,
+    R: Into<Route> + Display,
+{
+    let mut buf = Vec::new();
+    req.encode(&mut buf)?;
+    assert_request_match(schema, &buf);
+    trace! {
+        target = target,
+        id     = %req.header().id(),
+        method = ?req.header().method(),
+        path   = %req.header().path(),
+        body   = %req.header().has_body(),
+        "-> {label}"
+    };
+    let vec: Vec<u8> = ctx.send_and_receive(route, buf).await?;
+    Ok(vec)
+}
+
+/// Decode response header only, without processing the message body.
+pub(crate) fn is_ok(target: &str, label: &str, buf: &[u8]) -> ockam_core::Result<()> {
+    let mut d = Decoder::new(buf);
+    let res = response(target, label, &mut d)?;
+    assert_response_match(None, buf);
+    if res.status() == Some(Status::Ok) {
+        Ok(())
+    } else {
+        Err(error(target, label, &res, &mut d))
+    }
+}
+
+/// Decode response and body.
+pub(crate) fn decode<'a, 'b, T: Decode<'b, ()>>(
+    target: &'a str,
+    label: &'a str,
+    schema: impl Into<Option<&'a str>>,
+    buf: &'b [u8],
+) -> ockam_core::Result<T> {
+    let mut d = Decoder::new(buf);
+    let res = response(target, label, &mut d)?;
+    if res.status() == Some(Status::Ok) {
+        assert_response_match(schema, buf);
+        d.decode().map_err(|e| e.into())
+    } else {
+        Err(error(target, label, &res, &mut d))
+    }
+}
+
+/// Decode response and an optional body.
+pub(crate) fn decode_option<'a, 'b, T: Decode<'b, ()>>(
+    target: &'a str,
+    label: &'a str,
+    schema: impl Into<Option<&'a str>>,
+    buf: &'b [u8],
+) -> ockam_core::Result<Option<T>> {
+    let mut d = Decoder::new(buf);
+    let res = response(target, label, &mut d)?;
+    match res.status() {
+        Some(Status::Ok) => {
+            assert_response_match(schema, buf);
+            Ok(Some(d.decode()?))
+        }
+        Some(Status::NotFound) => Ok(None),
+        _ => Err(error(target, label, &res, &mut d)),
+    }
+}
+
+/// Decode and log response header.
+pub(crate) fn response(
+    target: &str,
+    label: &str,
+    dec: &mut Decoder<'_>,
+) -> ockam_core::Result<Response> {
+    let res: Response = dec.decode()?;
+    trace! {
+        target = target,
+        id     = %res.id(),
+        re     = %res.re(),
+        status = ?res.status(),
+        body   = %res.has_body(),
+        "<- {label}"
+    }
+    Ok(res)
```

Comment:
This might be helpful. For now I'll leave it as it is because I have to refactor more stuff in this crate. If this function doesn't need to be extended, I'll implement `Value` and get rid of the function.